File: src/test/java/Dl4jTest.java
import com.google.common.collect.Lists;
import com.oanda.predictor.domain.Candle;
import com.oanda.predictor.util.LSTMNetwork;
import com.oanda.predictor.util.StockDataSetIterator;
import com.opencsv.CSVReader;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartPanel;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.axis.NumberAxis;
import org.jfree.chart.plot.PlotOrientation;
import org.jfree.chart.plot.XYPlot;
import org.jfree.data.xy.XYSeries;
import org.jfree.data.xy.XYSeriesCollection;
import org.jfree.ui.ApplicationFrame;
import org.jfree.ui.RefineryUtilities;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.primitives.Pair;

import java.io.FileReader;
import java.io.IOException;
import java.util.List;

public class Dl4jTest {

    public static void main(String[] args) {
        StockDataSetIterator iterator = new StockDataSetIterator(getData(), 0.7);
        MultiLayerNetwork net = LSTMNetwork.buildLstmNetworks(iterator);

        List<Pair<INDArray, Double>> testData = iterator.getTest();
        List<Double> predicts = Lists.newArrayList();
        List<Double> actuals = Lists.newArrayList();

        testData.forEach(indArrayDoublePair -> {
            INDArray output = net.rnnTimeStep(indArrayDoublePair.getKey());
            predicts.add(StockDataSetIterator.deNormalize(
                    output.getDouble(0), iterator.getCloseMin(), iterator.getCloseMax()
            ));
            actuals.add(indArrayDoublePair.getValue());
        });

        plote(actuals, predicts);
    }

    private static void plote(List<Double> actuals, List<Double> predicts) {
        final XYSeries series = new XYSeries("USD/SEK");
        final XYSeries series2 = new XYSeries("USD/SEK PREDICT");

        double max = 0;
        double min = 999;
        for (int i = 0; i < actuals.size(); i++) {
            if (actuals.get(i) > max) max = actuals.get(i);
            if (actuals.get(i) < min) min = actuals.get(i);
            if (predicts.get(i) > max) max = predicts.get(i);
            if (predicts.get(i) < min) min = predicts.get(i);

            series.add(i, actuals.get(i));
            series2.add(i, predicts.get(i));
        }

        final XYSeriesCollection data = new XYSeriesCollection();
        data.addSeries(series);
        data.addSeries(series2);

        final JFreeChart chart = ChartFactory.createXYLineChart(
                "USD/SEK",
                "Tick M5",
                "ClosePrice",
                data,
                PlotOrientation.VERTICAL,
                true,
                true,
                false
        );

        NumberAxis range = (NumberAxis) ((XYPlot) chart.getPlot()).getRangeAxis();
        range.setRange(min, max);

        final ChartPanel chartPanel = new ChartPanel(chart);
        chartPanel.setPreferredSize(new java.awt.Dimension(1024, 600));

        ApplicationFrame appFrane = new ApplicationFrame("USD/SEK");
        appFrane.setContentPane(chartPanel);
        appFrane.pack();
        RefineryUtilities.centerFrameOnScreen(appFrane);
        appFrane.setVisible(true);
    }

    private static List<Candle> getData() {
        List<Candle> data = Lists.newArrayList();
        String csvFile = Dl4jTest.class.getResource("data.csv").getFile();
        try {
            CSVReader reader = new CSVReader(new FileReader(csvFile));
            String[] line;
            while ((line = reader.readNext()) != null) {
                Candle candle = new Candle();
                candle.setOpen(Double.valueOf(line[1]));
                candle.setHigh(Double.valueOf(line[2]));
                candle.setLow(Double.valueOf(line[3]));
                candle.setClose(Double.valueOf(line[4]));
                candle.setVolume(Double.valueOf(line[5]).intValue());
                data.add(candle);
            }
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }

        return data.subList(0, 1000);
    }
}

File: src/test/java/Dl4jTest.kt
import com.google.common.collect.Lists
import com.oanda.predictor.util.CSVUtil
import com.oanda.predictor.util.LSTMNetwork
import com.oanda.predictor.util.StockDataSetIterator
import org.deeplearning4j.util.ModelSerializer
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartUtilities
import org.jfree.chart.axis.NumberAxis
import org.jfree.chart.plot.PlotOrientation
import org.jfree.chart.plot.XYPlot
import org.jfree.data.xy.XYSeries
import org.jfree.data.xy.XYSeriesCollection
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileOutputStream
import java.io.IOException


class Dl4jExpApplicationTest

private val log = LoggerFactory.getLogger(Dl4jExpApplicationTest::class.java)
private const val DATA_SIZE = 5_000
private const val SPLIT_RATIO = 0.9

const val dataFileName = "Data.csv"
const val networkFileName = "NeuralNetwork"

fun main(args: Array<String>) {
    predictTest()
}

@Throws(IOException::class)
fun predictTest() {
    val dataFile = Dl4jExpApplicationTest::class.java.getResource("/$dataFileName").file
    val data = CSVUtil.getCandles(dataFile, DATA_SIZE)
    val iterator = StockDataSetIterator(data, SPLIT_RATIO)
    val neuralNetworkFile = Dl4jExpApplicationTest::class.java.getResource(networkFileName)
    val net = if (neuralNetworkFile != null && File(neuralNetworkFile.file).exists()) {
        try {
            ModelSerializer.restoreMultiLayerNetwork(neuralNetworkFile.file)
        } catch (ex: IOException) {
            log.error(ex.message)
            return
        }

    } else {
        LSTMNetwork.buildLstmNetworks(iterator)
    }

    val testData = iterator.test
    val predicts = Lists.newArrayList<Double>()
    val actually = Lists.newArrayList<Double>()
    val indicators: Array<DoubleArray>? = iterator.indicators

    testData.forEach { indArrayDoublePair ->
        val output = net.rnnTimeStep(indArrayDoublePair.key)
        if (indicators!!.size >= 6) {
            indicators[0][indicators[0].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(0), iterator.mins[0], iterator.maxs[0]
            )
            indicators[1][indicators[1].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(1), iterator.mins[1], iterator.maxs[1]
            )
            indicators[2][indicators[2].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(2), iterator.mins[2], iterator.maxs[2]
            )
            indicators[3][indicators[3].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(3), iterator.mins[3], iterator.maxs[3]
            )
            indicators[4][indicators[4].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(4), iterator.mins[4], iterator.maxs[4]
            )
            indicators[5][indicators[5].size - testData.size + predicts.size - 1] = StockDataSetIterator.deNormalize(
                    output.getDouble(5), iterator.mins[5], iterator.maxs[5]
            )
        }
        predicts.add(StockDataSetIterator.deNormalize(
                output.getDouble(6), iterator.closes[0], iterator.closes[1]
        ))
        actually.add(StockDataSetIterator.deNormalize(indArrayDoublePair.value, iterator.closes[0], iterator.closes[1]))
    }

    plote(actually, predicts, indicators)
}

private fun plote(actuals: List<Double>, predicts: List<Double>, indicators: Array<DoubleArray>?) {
    val series = XYSeries("actually")
    val series2 = XYSeries("predict")

    var max = 0.0
    var min = 999.0
    for (i in predicts.indices) {
        val actual = actuals[i]
        if (actual > max) max = actual
        if (actual < min) min = actual

        val predict = predicts[i]
        if (predict > max) max = predict
        if (predict < min) min = predict

        series.add(i.toDouble(), actual)
        series2.add(i.toDouble(), predict)
    }

    val data = XYSeriesCollection()
    data.addSeries(series)
    data.addSeries(series2)

    for ((k, indicator) in indicators!!.withIndex()) {
        val seriesInd = XYSeries(k)
        var indexStart = indicator.size - predicts.size
        if (indexStart < 0) indexStart = 0
        var i = 0
        while (i < actuals.size) {
            seriesInd.add(i.toDouble(), indicator[indexStart + i])
            i++
        }
        data.addSeries(seriesInd)
    }

    val chart = ChartFactory.createXYLineChart(
            "Symbol",
            "Ticks",
            "ClosePrice",
            data,
            PlotOrientation.VERTICAL,
            true,
            true,
            false
    )

    val range = (chart.plot as XYPlot).rangeAxis as NumberAxis
    range.setRange(min, max)

    try {
        val out = FileOutputStream("out/Symbol_$DATA_SIZE.png")
        ChartUtilities.writeChartAsPNG(out, chart, 1024, 600)
    } catch (ex: IOException) {
        log.error(ex.message)
    }
}
